Firestore/core/src/firebase/firestore/nanopb/message.h

```diff
@@ -0,0 +1,177 @@
+#include "Firestore/core/src/firebase/firestore/remote/serializer.h"
+#include "Firestore/core/src/firebase/firestore/util/hard_assert.h"
+#include "Firestore/core/src/firebase/firestore/util/statusor.h"
+#include "grpcpp/support/byte_buffer.h"
+
+namespace firebase {
+namespace firestore {
+
+namespace remote {
+namespace bridge {
+
+class DatastoreSerializer;
+class WatchStreamSerializer;
+class WriteStreamSerializer;
+
+}  // namespace bridge
+}  // namespace remote
+
+namespace nanopb {
+
+template <typename T>
+class Message;
+
+template <typename T>
+using MaybeMessage = util::StatusOr<Message<T>>;
+
+/**
+ * A unique-ownership RAII wrapper for Nanopb-generated message types.
+ *
+ * Nanopb-generated message types are plain C structs that contain some
+ * dynamically-allocated memory and should be deallocated by calling
+ * `pb_release`; `Message` implements a simple RAII wrapper that does just that.
+ * For simplicity, `Message` implements unique ownership and is immutable after
+ * construction (not counting friend classes). Use `proto()` member function to
+ * access the underlying proto.
+ *
+ * Accessing a moved-from instance will result in an assertion failure.
+ *
+ * Note that moving *isn't* a particularly cheap operation in the general case.
+ * Even without doing deep copies, Nanopb-generated messages may contain *a lot*
+ * of member variables.
+ */
+template <typename T>
+class Message {
+ public:
+  /**
+   * Attempts to parse a Nanopb message from the given `byte_buffer`.
+   *
+   * If the given bytes are ill-formed, returns a failed `Status`.
+   *
+   * `fields` is the Nanopb-generated descriptor of message `T`, which are named
+   * by adding a `_fields` suffix to the name of the message. E.g., for
+   * `google_firestore_v1_Foo` message, the corresponding fields descriptor will
+   * be named `google_firestore_v1_Foo_fields`.
+   */
+  static MaybeMessage<T> Parse(const pb_field_t* fields,
+                               const grpc::ByteBuffer& byte_buffer);
+
+  ~Message() {
+    if (owns_proto()) {
+      remote::Serializer::FreeNanopbMessage(fields_, &proto_);
+    }
+  }
+
+  Message(const Message&) = delete;
+  Message& operator=(const Message&) = delete;
+
+  Message(Message&& other) noexcept
+      : fields_{other.fields_}, proto_{other.proto_} {
+    other.fields_ = nullptr;
+  }
+
+  Message& operator=(Message&& other) noexcept {
+    fields_ = other.fields_;
+    proto_ = other.proto_;
+    other.fields_ = nullptr;
+  }
+
+  const T& proto() const {
+    HARD_ASSERT(owns_proto(),
+                "Attempted to access a moved-from nanopb::Message");
+    return proto_;
+  }
+
+  /**
+   * Serializes this message into a byte buffer.
+   *
+   * The lifetime of the return value is entirely independent of this message.
+   */
+  grpc::ByteBuffer CreateByteBuffer() const;
+
+ private:
+  // For access to `mutable_proto` and the explicit constructor. Most code
```

Comment:
This made the class's public interface nicer. Let me know if you have any objections or suggestions.